File: reels-to-capcut/reels_to_capcut/vision_ocr.swift
import CoreGraphics
import Foundation
import ImageIO
import Vision

struct Response: Codable {
    let lines: [String]?
    let error: String?
}

func emit(_ response: Response, code: Int32 = 0) -> Never {
    let data = try! JSONEncoder().encode(response)
    FileHandle.standardOutput.write(data)
    FileHandle.standardOutput.write("\n".data(using: .utf8)!)
    exit(code)
}

guard CommandLine.arguments.count > 1 else {
    emit(Response(lines: nil, error: "image path required"), code: 2)
}

let path = CommandLine.arguments[1]
guard let source = CGImageSourceCreateWithURL(URL(fileURLWithPath: path) as CFURL, nil),
      let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
    emit(Response(lines: nil, error: "cannot read image"), code: 3)
}

let request = VNRecognizeTextRequest()
request.recognitionLevel = .accurate
request.recognitionLanguages = ["ko-KR", "en-US"]
request.usesLanguageCorrection = true

do {
    try VNImageRequestHandler(cgImage: image, options: [:]).perform([request])
    let ordered = (request.results ?? []).sorted {
        if abs($0.boundingBox.midY - $1.boundingBox.midY) > 0.03 {
            return $0.boundingBox.midY > $1.boundingBox.midY
        }
        return $0.boundingBox.minX < $1.boundingBox.minX
    }
    let lines = ordered.compactMap { $0.topCandidates(1).first?.string }
    emit(Response(lines: lines, error: nil))
} catch {
    emit(Response(lines: nil, error: error.localizedDescription), code: 4)
}

File: reels-to-capcut/reels_to_capcut/vision_ocr.py
#!/usr/bin/env /usr/bin/python3
"""Small macOS Vision bridge, intentionally run with Apple's system Python."""

from __future__ import annotations

import json
import sys


def recognize(path: str) -> list[str]:
    from Foundation import NSURL
    from Quartz import CGImageSourceCreateImageAtIndex, CGImageSourceCreateWithURL
    import Vision

    source = CGImageSourceCreateWithURL(NSURL.fileURLWithPath_(path), None)
    image = CGImageSourceCreateImageAtIndex(source, 0, None)
    request = Vision.VNRecognizeTextRequest.alloc().init()
    request.setRecognitionLevel_(Vision.VNRequestTextRecognitionLevelAccurate)
    request.setRecognitionLanguages_(["ko-KR", "en-US"])
    request.setUsesLanguageCorrection_(True)
    handler = Vision.VNImageRequestHandler.alloc().initWithCGImage_options_(image, {})
    ok, error = handler.performRequests_error_([request], None)
    if not ok:
        raise RuntimeError(str(error))
    found: list[tuple[float, float, str]] = []
    for observation in request.results() or []:
        candidates = observation.topCandidates_(1)
        if not candidates:
            continue
        box = observation.boundingBox()
        found.append((-float(box.origin.y), float(box.origin.x), str(candidates[0].string())))
    return [text for _, _, text in sorted(found) if text.strip()]


if __name__ == "__main__":
    try:
        print(json.dumps({"lines": recognize(sys.argv[1])}, ensure_ascii=False))
    except Exception as exc:
        print(json.dumps({"error": str(exc)}, ensure_ascii=False))
        raise SystemExit(1)
